server/src/main/java/org/opensearch/common/metrics/ResourceTracker.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.metrics;
+
+import com.sun.management.ThreadMXBean;
+import org.opensearch.common.SuppressForbidden;
+
+import java.lang.management.ManagementFactory;
+
+/*
+    Simple tracker for CPU consumption and memory allocated by the current thread
+*/
+@SuppressForbidden(reason = "ThreadMXBean enables tracking resource consumption by a thread. "
+    + "It is platform dependent and i am not aware of an alternate mechanism to extract this info")
+public class ResourceTracker {
```

Comment:
I think instead of having `ResourceTracker` in `SearchContext`, we can move it as interface to Task class itself. This interface can then expose the methods like:
1) `setup/initialize` where it can initialize the current values from MXBean.
2) `recordStats()` where it can record the diff

Then once `setTask` is called on Context object, it can initialize the collector/tracker. In `updateResourceTracking` it can call the `context.getResourceTracker().recordStats()`. This will avoid the need for `reset` as each task will be updating its own internal stats. Context object will be set with new task for query and fetch phase separately.

Also how about renaming `ResourceTracker` as `StatsCollector` or `MetricsCollector` ?

Reply:
> moving `ResourceTracker` to `Task`
* The reason I kept it in `SearchContext` is, `ResourceTracker` with its thread level stats is unique to tracking resource consumption for queries only. Currently, I am not seeing other consumers. We can move it upwards if there are other `Task` that would need similar logic
* Also, if the concurrent segment search becomes the default in future, we might need multiple `ResourceTracker` per search task
* I do see your point to avoid/optimize `reset` - and need to see the scenarios where the context could be reused. Either way will check if memory/cpu is non zero as a condition on pre-empting reset
* what are your thoughts?

> Also how about renaming ResourceTracker as StatsCollector or MetricsCollector ?
* `Stats` has a special meaning as `stats` it is a [request param](https://www.elastic.co/guide/en/elasticsearch/reference/7.10/search-search.html) that allows grouping stats on a tag
* `Collector` has a special meaning too - [Lucene collector](https://lucene.apache.org/core/8_0_0/core/org/apache/lucene/search/Collector.html)
* What are your thoughts on 
   * `MetricsTracker` - tracking metrics - resource consumption, other future metrics
   * `ResourceTracker` - tracking resource consumption

Reply:
> moving `ResourceTracker` to `Task`

- The reason for proposing `ResourceTracker` to move it to `Task` is, these metrics/stats belongs to `task`. Putting it within context is creating the need for `reset` since context is used by multiple tasks. Considering `Task` will be tracking metrics then having an interface which provides the API to `update/get` stats related information in Task will fit well. I see that not all tasks is currently supporting the metrics collection, however that can be achieved by implementing the  `update/get` only by specific search related tasks. The base task can be a no-op for these calls.
- For concurrent segment search, there still will be single task and context for search but segment level operation will be done in parallel by multiple threads. So each of these threads will now need to update the stats for same task.

> Also how about renaming ResourceTracker as StatsCollector or MetricsCollector ?

I don't think `stats/collector` definition in the example you shared makes it reserved word for only those modules. It is a general term and can be used in multiple places, for example: `SearchExecutionStatsCollector`. But I am fine with `MetricsTracker` as well since that will cover other categories like latencies too and not just resources.

Reply:
have pushed a new revision which incorporates this feedback:
1. Got rid of the `reset` calls as `SearchContext` is not reused anymore 
2. Renamed to MetricsTracker